### PR TITLE
add pipe class and test

### PR DESCRIPTION
**Pull request by zs0930**

### CODE REVIEW

File: include/pipe.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#ifndef PIPE_NETWORK_PIPE_H_
+#define PIPE_NETWORK_PIPE_H_
+
+#include "node.h"
```
Comment (kks32) [inline include/pipe.h]: This `node.h` should be after  C++ includes and c-includes

File: include/pipe.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#include <array>
+#include <math.h>
```
Comment (kks32) [inline include/pipe.h]: use `cmath`

File: include/pipe.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  //! Return pipe broken status
+  //! \retval isbroken_ pipe broken status
+  bool isbroken() const { return isbroken_; }
+
+ private:
+  //! pipe id
+  unsigned id_{std::numeric_limits<unsigned>::max()};
+  //! array of node pointers which form the pipe
+  std::array<std::shared_ptr<Node>, 2> array_node_ptr_;
```
Comment (kks32) [inline include/pipe.h]: Rename to `nodes_`

File: include/pipe.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  //! Move constructor
+  Pipe(Pipe&&) = delete;
+
+  //! Return id
+  //! \retval id_ id of the pipe
+  unsigned id() const { return id_; }
+
+  //! Return array of node pointers
+  //! \retval array_node_ptr_ array of node pointers
+  std::array<std::shared_ptr<Node>, 2> array_node_ptr() const {
```
Comment (kks32) [inline include/pipe.h]: You shouldn't return nodes this is giving up control. So philosophically, this shouldn't happen. Pipe should own the node objects, you can't pass it to something else.

File: tests/pipe_test.cc [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// Pipe test
+#include "node.h"
+#include "pipe.h"
+
+#define CATCH_CONFIG_MAIN
+#include "catch.hpp"
+
+//! \brief Ckeck pipe class
```
Comment (kks32) [inline tests/pipe_test.cc]: Check your spelling for check ;)

File: tests/pipe_test.cc [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+TEST_CASE("Pipe is checked", "[Pipe]") {
+  unsigned nodeid1 = 100;
```
Comment (kks32) [inline tests/pipe_test.cc]: See node_test.cc for comments.

File: include/pipe.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  //! \retval discharge_ discharge
+  unsigned discharge() const { return discharge_; }
+
+  //! Assign radius of the pipe
+  //! \param[in] radius radius of the pipe
+  void radius(double radius) { radius_ = radius; }
+
+  //! Assign length of the pipe
+  //! \param[in] length of the pipe
+  // void length(double length) { length_ = length; }
```
Comment (kks32) [inline include/pipe.h]: You shouldn't assign the length of the pipe, rather calculate it. You have all the necessary information to calculate the length of the pipe

**Review by kks32 — CHANGES_REQUESTED** [review]

File: tests/pipe_test.cc (hunk 5 above)
Comment (kks32) [reply tests/pipe_test.cc]: It's nor brief here.

File: tests/pipe_test.cc [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  REQUIRE(pipe->id() == pipeid);
+  //! Check array of node pointers
+  REQUIRE(pipe->array_node_ptr()[0]->id() == nodeid1);
+  REQUIRE(pipe->array_node_ptr()[1]->id() == nodeid2);
+  //! Check pipe broken status and initialized status
+  REQUIRE(pipe->isbroken() == false);
+
+  //! Check radius, discharge and max flow velocity of the pipe
+  SECTION("check radius of the pipe") {
+    double radius = 10.0;
```
Comment (kks32) [inline tests/pipe_test.cc]: If things won't change use `const`

File: tests/pipe_test.cc [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    double max_velocity = 100.0;
+    double discharge = M_PI * 1.e4;
```
Comment (kks32) [inline tests/pipe_test.cc]: What's `1.e-4`?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: tests/pipe_test.cc (hunk 9 above)
Comment (zs0930) [reply tests/pipe_test.cc]: 1.e4
to calculate the max discharge using radius and max velocity, 10^2*100

File: include/pipe.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  // }
+
+  //! Assign maximum allowable velocity
+  //! \param[in] max_flowrate maximum allowable velocity of flow in the pipe
+  void max_velocity(double max_velocity) { max_velocity_ = max_velocity; }
+
+  //! calculate maximum allowable discharge based on ridius and maximum
+  //! allowable velocity \retval max_discharge maximun allowable discharge
+  double max_discharge() {
+    double max_discharge = max_velocity_ * M_PI * pow(radius_, 2);
```
Comment (kks32) [inline include/pipe.h]: `return (max_velocity_ * M_PI * pow(radius_, 2));`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/pipe.h [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  //! Move constructor
+  Pipe(Pipe&&) = delete;
+
+  //! Return id
+  //! \retval id_ id of the pipe
+  unsigned id() const { return id_; }
+
+  //! Return discharge in the pipe
+  //! \retval discharge_ discharge
+  unsigned discharge() const { return discharge_; }
```
Comment (kks32) [inline include/pipe.h]: Discharge is not calculated, how are we returning this?

File: tests/pipe_test.cc [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+#define CATCH_CONFIG_MAIN
+#include "catch.hpp"
+
+// Check pipe class
+TEST_CASE("Pipe is checked", "[Pipe]") {
+  // Tolerance
+  const double tolerance = 1.e-12;
+
+  // Node index
```
Comment (kks32) [inline tests/pipe_test.cc]: Node indices

File: tests/pipe_test.cc [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  const unsigned nodeid1 = 100;
+  const unsigned nodeid2 = 101;
+
+  // Node coordinates
```
Comment (kks32) [inline tests/pipe_test.cc]: Nodal coordinates

File: tests/pipe_test.cc [hunk 14]
```diff
@@ -0,0 +1,55 @@
+// Pipe test
+#include "node.h"
+#include "pipe.h"
+
+#define CATCH_CONFIG_MAIN
```
Comment (kks32) [inline tests/pipe_test.cc]: You don't need this.

File: tests/pipe_test.cc [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+#include "catch.hpp"
```
Comment (kks32) [inline tests/pipe_test.cc]: Order of includes means this should be at the top

File: tests/pipe_test.cc [hunk 16]
```diff
@@ -0,0 +1,55 @@
+// Pipe test
+#include "node.h"
```
Comment (kks32) [inline tests/pipe_test.cc]: After `catch.hpp`

File: tests/pipe_test.cc [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  const unsigned nodeid1 = 100;
+  const unsigned nodeid2 = 101;
+
+  // Node coordinates
+  const Eigen::Vector3d coords1(10.0, 20.0, 30.0);
+  const Eigen::Vector3d coords2(11.1, 21.1, 31.1);
+
+  // Creat an array of two node pointers with previous defined index and
+  // coordinates of the nodes
+  std::array<std::shared_ptr<Node>, 2> nodeptr;
```
Comment (kks32) [inline tests/pipe_test.cc]: `nodes` instead of `nodeptr`, don't include the type as the type might change.

File: tests/pipe_test.cc [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  nodeptr[0] = std::make_shared<Node>(nodeid1, coords1);
+  nodeptr[1] = std::make_shared<Node>(nodeid2, coords2);
+
+  // Pipe index
+  const unsigned pipeid = 200;
+
+  // Creat a pipe based on previous created node pointers
+  auto pipe(std::make_unique<Pipe>(pipeid, nodeptr));
```
Comment (kks32) [inline tests/pipe_test.cc]: auto pipe = std::make_unique<Pipe>(pipeid, nodeptr);

File: include/pipe.h [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+//! Pipe class
+//! \brief Class that stores the information about pipes
+class Pipe {
+
+ public:
+  // Constructor with id and node pointers
+  //! \param[in] id node id
+  //! \param[in] nodes array of node pointers
+  Pipe(unsigned id, const std::array<std::shared_ptr<Node>, 2>& nodes)
```
Comment (kks32) [inline include/pipe.h]: Move this to a source file

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/pipe.cc [hunk 20]
```diff
@@ -0,0 +1,24 @@
+// Constructor with id and node pointers
+//! \param[in] id node id
+//! \param[in] nodes array of node pointers
+Pipe::Pipe(unsigned id, const std::array<std::shared_ptr<Node>, 2>& nodes)
+    : id_{id}, nodes_{nodes} {
+  Eigen::Vector3d distance =
```
Comment (kks32) [inline include/pipe.cc]: do it in one step. `length_ = (...).norm()`

File: include/pipe.cc [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+      nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates();
+  length_ = distance.norm();
+}
+
+//! Calculate and return discharge using Darcy-Weisbach equation
+//! retval discharge_ discharge in the pipe
+double Pipe::discharge() {
+  if (nodes_[0]->ishead() && nodes_[1]->ishead()) {
+    double dhead = abs(nodes_[0]->head() - nodes_[1]->head());
```
Comment (kks32) [inline include/pipe.cc]: do we need to use `std::fabs`?

Comment (kks32) [inline include/pipe.cc]: Please use const, when things won't change.

File: include/pipe.cc [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    discharge_ = sqrt(dhead * pow(M_PI, 2) * g_ * pow(2 * radius_, 5) /
```
Comment (kks32) [inline include/pipe.cc]: Show the equation.

File: include/pipe.cc [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+//! \param[in] nodes array of node pointers
+Pipe::Pipe(unsigned id, const std::array<std::shared_ptr<Node>, 2>& nodes)
+    : id_{id}, nodes_{nodes} {
+  Eigen::Vector3d distance =
+      nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates();
+  length_ = distance.norm();
+}
+
+//! Calculate and return discharge using Darcy-Weisbach equation
+//! retval discharge_ discharge in the pipe
```
Comment (kks32) [inline include/pipe.cc]: You don't need this in source files

File: include/pipe.cc [hunk 24]
```diff
@@ -0,0 +1,24 @@
+// Constructor with id and node pointers
+//! \param[in] id node id
```
Comment (kks32) [inline include/pipe.cc]: No need to repeat in source files

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/pipe.cc [hunk 25]
```diff
@@ -1,20 +1,18 @@
...
 }
 
 //! Calculate and return discharge using Darcy-Weisbach equation
-//! retval discharge_ discharge in the pipe
 double Pipe::discharge() {
   if (nodes_[0]->ishead() && nodes_[1]->ishead()) {
-    double dhead = abs(nodes_[0]->head() - nodes_[1]->head());
-    discharge_ = sqrt(dhead * pow(M_PI, 2) * g_ * pow(2 * radius_, 5) /
+    const double dhead = nodes_[0]->head() - nodes_[1]->head();
+    discharge_ = sqrt(abs(dhead) * pow(M_PI, 2) * g_ * pow(2 * radius_, 5) /
```
Comment (kks32) [inline include/pipe.cc]: Please use std::abs or std::fabs

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

File: include/mesh.h [hunk 26]
```diff
@@ -0,0 +1,110 @@
+#ifndef PIPE_NETWORK_MESH_H_
```
Comment (kks32) [inline include/mesh.h]: Could you please remove mesh from your pipe class Pull Request.

File: include/pipe.cc [hunk 27]
```diff
@@ -1,15 +1,17 @@
 // Constructor with id and node pointers
 Pipe::Pipe(unsigned id, const std::array<std::shared_ptr<Node>, 2>& nodes)
     : id_{id}, nodes_{nodes} {
+  for (unsigned i = 0; i < end_coordinates_.size(); ++i)
+    end_coordinates_.at(i) = nodes_.at(i)->coordinates();
   length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
 }
 
 //! Calculate and return discharge using Darcy-Weisbach equation
 double Pipe::discharge() {
```
Comment (kks32) [inline include/pipe.cc]: Add equation

File: include/pipe.cc [hunk 28]
```diff
@@ -20,3 +22,12 @@ double Pipe::discharge() {
         "equation");
   }
 }
+
+//! Check whether the pipe is constructed based on a given node
+bool Pipe::isnode(std::shared_ptr<Node> node) {
```
Comment (kks32) [inline include/pipe.cc]: Pass by const reference

Comment (kks32) [inline include/pipe.cc]: what is `isnode`?

Comment (kks32) [reply include/pipe.cc]: Why do we need it?

File: include/pipe.h [hunk 29]
```diff
@@ -35,6 +35,12 @@ class Pipe {
   //! \retval id_ id of the pipe
   unsigned id() const { return id_; }
 
+  //! Return coordinates at two ends of the pipe
```
Comment (kks32) [inline include/pipe.h]: Why do we need this function?

Comment (kks32) [reply include/pipe.h]: Where is it calculated?

File: tests/mesh_test.cc [hunk 30]
```diff
@@ -0,0 +1,74 @@
+#include "catch.hpp"
```
Comment (kks32) [inline tests/mesh_test.cc]: Remove mesh testing from this PR

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/pipe.h [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+  //! length of the pipe
+  double length_;
+  //! Darcy friction factor of the pipe
+  double darcy_friction_{std::numeric_limits<double>::max()};
+  //! maximum allowable velocity of flow in the pipe
+  double max_velocity_{std::numeric_limits<double>::max()};
+  //! whether the pipe is broken
+  bool isbroken_{false};
+  //! gravitational acceleration
+  const double g_{9.81};
```
Comment (kks32) [inline include/pipe.h]: Why is gravity defined per pipe? and why only one coordinate of acceleration?

File: tests/pipe_test.cc [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+    const double max_discharge = M_PI * 1.e4;
+    const double head1 = 110.0;
+    const double head2 = 100.0;
+    const double darcy_friction = 0.1;
+    const double discharge =
+        sqrt(10 * pow(M_PI, 2) * 9.81 * pow(2 * 10, 5) / (8 * 0.1));
+    nodes[0]->head(head1);
+    nodes[1]->head(head2);
+    pipe->radius(radius);
+    pipe->max_velocity(max_velocity);
```
Comment (kks32) [inline tests/pipe_test.cc]: Could you write tests for these 3 functions?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/node.h [hunk 33]
```diff
@@ -3,9 +3,13 @@
 
 #include <Eigen/Dense>
 
+namespace pipenetwork {
+class Node;
+}
+
 //! Node class
```
Comment (kks32) [inline include/node.h]: You can encapsulate the whole class in a namespace, no need to predefine.

File: include/pipe.h [hunk 34]
```diff
@@ -8,16 +8,22 @@
...
+#include "settings.h"
+
+namespace pipenetwork {
+class Pipe;
+}
 
 //! Pipe class
 //! \brief Class that stores the information about pipes
-class Pipe {
+class pipenetwork::Pipe {
```
Comment (kks32) [inline include/pipe.h]: Write the whole class within the namespace.

File: include/pipe.h [hunk 35]
```diff
@@ -52,7 +58,7 @@ class Pipe {
   //! Calculate and return discharge using Darcy-Weisbach equation:
   //! dhead = (8*darcy_factor*pow(discharge,2)/(pow(M_PI,2)*g*pow(2*radius,5))
   //! That is, discharge =
-  //! sqrt(dhead*pow(M_PI,2)*g*pow(2*radius,5)/(8*darcy_friction)); retval
+  //! sqrt(dhead*pow(M_PI,2)*g*pow(2*radius,5)/(8*darcy_friction)); \retval
```
Comment (kks32) [inline include/pipe.h]: Remove `\retval`

File: include/settings.h [hunk 36]
```diff
@@ -0,0 +1,9 @@
+#ifndef PIPE_NETWORK_SETTING_H_
+#define PIPE_NETWORK_SETTING_H_
+
+namespace pipenetwork {
+//! define gravitational acceleration in x, y, z direction
+const Eigen::Vector3d g(0.00, 0.00, 9.81);
```
Comment (kks32) [inline include/settings.h]: Never use a single letter character, use `Gravity` instead.

File: include/pipe.cc [hunk 37]
```diff
@@ -0,0 +1,23 @@
+// Constructor with id and node pointers
+pipenetwork::Pipe::Pipe(
+    unsigned id, const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes)
+    : id_{id}, nodes_{nodes} {
+  length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
+}
+
+//! Calculate and return discharge using Darcy-Weisbach equation
```
Comment (kks32) [inline include/pipe.cc]: Remove `!`

File: include/pipe.h [hunk 38]
```diff
@@ -0,0 +1,96 @@
...
+namespace pipenetwork {
+class Pipe;
+}
+
+//! Pipe class
+//! \brief Class that stores the information about pipes
+class pipenetwork::Pipe {
+
+ public:
+  // Constructor with id and node pointers
```
Comment (kks32) [inline include/pipe.h]: Add `!` to //

File: include/pipe.h [hunk 39]
```diff
@@ -0,0 +1,96 @@
...
+  //! \param[in] id node id
```
Comment (kks32) [inline include/pipe.h]: Node id or pipe id?

File: include/pipe.h [hunk 40]
```diff
@@ -0,0 +1,96 @@
...
+  //! Assign Dracy friction factor
+  //! \param[in] darcy_friction Darcy friction factor of the pipe
+  void darcy_friction(double darcy_friction) {
+    darcy_friction_ = darcy_friction;
+  }
+
+  //! Calculate and return discharge using Darcy-Weisbach equation:
+  //! dhead = (8*darcy_factor*pow(discharge,2)/(pow(M_PI,2)*g*pow(2*radius,5))
+  //! That is, discharge =
+  //! sqrt(dhead*pow(M_PI,2)*g*pow(2*radius,5)/(8*darcy_friction)); \retval
```
Comment (kks32) [inline include/pipe.h]: Fix `\retval` to next line.

File: include/pipe.h [hunk 41]
```diff
@@ -0,0 +1,96 @@
...
+  //! discharge_ discharge in the pipe
```
Comment (kks32) [inline include/pipe.h]: `//! \retval`

Comment (kks32) [reply include/pipe.h]: Add what the function does

File: include/settings.h (hunk 36 above)
Comment (kks32) [inline include/settings.h]: Could you please show me where you found this way of initialisation? using () for initilisation?

File: tests/pipe_test.cc [hunk 42]
```diff
@@ -0,0 +1,68 @@
...
+  REQUIRE(pipe->isbroken() == false);
+
+  // Check radius, discharge, max flow velocity and Darcy friction factor of the
+  // pipe
+  SECTION(
+      "Check radius, discharge, max flow velocity and Darcy friction factor of "
+      "the pipe") {
+    const double radius = 10.0;
+    const double max_velocity = 100.0;
+    const double max_discharge = M_PI * 1.e4;
```
Comment (kks32) [inline tests/pipe_test.cc]: Why is it Pi * 1E-4?

File: tests/pipe_test.cc [hunk 43]
```diff
@@ -0,0 +1,68 @@
...
+
+  // Check pipe broken status and initialized status
+  REQUIRE(pipe->isbroken() == false);
+
+  // Check radius, discharge, max flow velocity and Darcy friction factor of the
+  // pipe
+  SECTION(
+      "Check radius, discharge, max flow velocity and Darcy friction factor of "
+      "the pipe") {
+    const double radius = 10.0;
```
Comment (kks32) [inline tests/pipe_test.cc]: Please add comments on units?

File: tests/pipe_test.cc [hunk 44]
```diff
@@ -0,0 +1,68 @@
...
+  nodes[1] = std::make_shared<pipenetwork::Node>(nodeid2, coords2);
+
+  // Pipe index
+  const unsigned pipeid = 200;
+
+  // Creat a pipe based on previous created node pointers
+  auto pipe = std::make_unique<pipenetwork::Pipe>(pipeid, nodes);
+
+  // Pipe length
+  const double length = sqrt(3 * pow(1.1, 2));
```
Comment (kks32) [inline tests/pipe_test.cc]: Explain why it is this way?

Comment (kks32) [reply tests/pipe_test.cc]: Add comment for  formula used.

File: tests/pipe_test.cc [hunk 45]
```diff
@@ -0,0 +1,68 @@
...
+  const unsigned nodeid2 = 101;
+
+  // Nodal coordinates
+  const Eigen::Vector3d coords1(10.0, 20.0, 30.0);
+  const Eigen::Vector3d coords2(11.1, 21.1, 31.1);
+
+  // Creat an array of two node pointers with previous defined index and
+  // coordinates of the nodes
+  std::array<std::shared_ptr<pipenetwork::Node>, 2> nodes;
+  nodes[0] = std::make_shared<pipenetwork::Node>(nodeid1, coords1);
```
Comment (kks32) [inline tests/pipe_test.cc]: Always use .at() instead of []

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/settings.h (hunk 36 above)
Comment (zs0930) [reply include/settings.h]: https://eigen.tuxfamily.org/dox/group__TutorialMatrixArithmetic.html

File: include/pipe.h (hunk 41 above)
Comment (zs0930) [reply include/pipe.h]: What the function does was stated at the first line of this part of comment

File: include/pipe.cc [hunk 46]
```diff
@@ -0,0 +1,23 @@
+// Constructor with id and node pointers
+pipenetwork::Pipe::Pipe(
+    unsigned id, const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes)
+    : id_{id}, nodes_{nodes} {
+  length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
+}
+
+// Calculate and return discharge using Darcy-Weisbach equation
+double pipenetwork::Pipe::discharge() {
+  if (nodes_[0]->ishead() && nodes_[1]->ishead()) {
```
Comment (kks32) [inline include/pipe.cc]: Could you add comments to say what you are doing?

File: include/pipe.cc [hunk 47]
```diff
@@ -0,0 +1,23 @@
...
+    const double dhead = nodes_[0]->head() - nodes_[1]->head();
+    discharge_ = sqrt(std::abs(dhead) * pow(M_PI, 2) * pipenetwork::Gravity(2) *
+                      pow(2 * radius_, 5) / (8 * darcy_friction_));
+    if (dhead < 0) {
+      discharge_ = -1 * discharge_;
+    }
+    return discharge_;
```
Comment (kks32) [inline include/pipe.cc]: You are returning before throw, always return at the end.

File: include/pipe.cc [hunk 48]
```diff
@@ -0,0 +1,23 @@
...
+  length_ = (nodes_.at(0)->coordinates() - nodes_.at(1)->coordinates()).norm();
+}
+
+// Calculate and return discharge using Darcy-Weisbach equation
+double pipenetwork::Pipe::discharge() {
+  if (nodes_[0]->ishead() && nodes_[1]->ishead()) {
+    const double dhead = nodes_[0]->head() - nodes_[1]->head();
+    discharge_ = sqrt(std::abs(dhead) * pow(M_PI, 2) * pipenetwork::Gravity(2) *
+                      pow(2 * radius_, 5) / (8 * darcy_friction_));
+    if (dhead < 0) {
```
Comment (kks32) [inline include/pipe.cc]: remove `{` and `}`

File: include/pipe.cc [hunk 49]
```diff
@@ -0,0 +1,23 @@
...
+      discharge_ = -1 * discharge_;
```
Comment (kks32) [inline include/pipe.cc]: `discharge_ *= -1.;`

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**